### PR TITLE
Feature/handle disabled code actions

### PR DESCRIPTION
**Pull request by nbfalcon**
- Handle disabled code actions (not shown in `lsp-execute-code-action` unless C-u prefix specified)
- Properly propertize preferred code actions
- Add `lsp-auto-fix`, which executes one of the preferred code actions
- Add `lsp-execute-code-action-by-type`, which executes a code action by given base type (e.g. "refactor", ...)

I have tested this a bit with `jdtls`, and this PR seems to not introduce a regression. However, I didn't have the chance to test the new disabled and preferred code action handling yet, which is why I am marking as draft.

There will be a corresponding `lsp-ui` PR soon, too (there is just one minor change I need to make, and to test this).

### CODE REVIEW

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:4890]: Ideally we would want the default code action to be preselected in the select dialog. If you don't want to adress it in this PR you may leave a TODO.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:4903]: If you are going to change it, then change it to pcase.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5251]: If I call explicitly a refactoring and the method is disabled I would like to see the reason for which it is disabled. E. g. if I call "Extract method" I might see an error "You have selected a range that cannot be extracted to a method".

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5289]: this seems like should be part of `lsp-execute-code-action`. E. g. if lsp-ui calls the method with disabled action it should get that error.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5302]: seems like if you drop lsp--select-enabled-action/lsp--select-action-enabled-or-die and pass an optional parameter to lsp-code-actions-at-point the code will be simplified/shortened a lot.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:4896]: I have implemented this. That feature was surprisingly easy to implement, requiring only a small change to `lsp-completing-read` and one to `lsp--select-action`.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Note: this is now an `error` instead of a `user-error` (reasoning in the commit message).

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed. There might be a prompt with the reasons listed individually, depending on the number of actions and `lsp-auto-execute-action`.

**Comment by nbfalcon** [comment]
Should I prepend "Review: " to the commit subject if it doesn't make it go over 50 lines?

**Comment by yyoncho** [comment]
> Should I prepend "Review: " to the commit subject if it doesn't make it go over 50 lines?

We dont have such a requirement. It will be squashed in the end so it shouldn't matter.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5246]: I guess this section is on the edge of being too trickier.

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5261]: can you elaborate on why we need this method?

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5252]: don't we need to filter the disabled and only in case there is no enabled action use the disabled one(s) error message?

**Review by yyoncho — COMMENTED** [review]
Looks good. I would have approved it if it was not marked as draft.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5261]: If there is some specific code action to be executed, e.g. `source.organizeImports`, `lsp-execute-code-action-by-kind` is enough. However, sometimes it can be useful to bring up all quickfixes, or all refactorings: in this case, `lsp-execute-code-action-by-kind` doesn't work, because it must match the kind exactly. My new function matches a base kind: `lsp-execute-code-action-by-type` "refactor" would match "refactor.move", "refactor.*", ... (tested with `jdtls`). This corresponds to the spec's "if the user requests a more specific type of code action, e.g. refactoring".

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5246]: What I had in mind here was the following: currently there is a single filter: ENABLED. In the future, more filters could be added, and using `--doto` with `cl-callf` allows adding them easily (e.g. `is-preferred`). I could rewrite this in terms of `if` if you consider that to be a large enough readability gain:
```emacs-lisp
  (let ((actions (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind))))
    (if enabled
        (seq-remove #'lsp:code-action-disabled? actions)
      acitons))
```

I just realized that the code has a bug: `lsp-code-action-disabled?` should be `lsp:code-action-disabled?`.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed.

**Comment by nbfalcon** [comment]
Now all that is left is to test this.

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el:5246]: I guess it could stay as it is, just it is a bit unusual. Once `-cond->>` lands in `dash` we can use it here(and few other places).

Comment (nbfalcon) [reply lsp-mode.el:5246]: `-cond->>` would certainly be awesome, and I remember that I wanted to use it somewhere but couldn't. However, I don't see how this would help here? The point of this extensible function style is to allow filters to be chained: e.g. first filter out all disabled actions, then those that don't match the type base, .... (we don't have that yet, but might in the future).

**Comment by nbfalcon** [comment]
As a side note, why not directly wait for `lsp-response-timeout` [here](https://github.com/emacs-lsp/lsp-mode/blob/0349a1cc0976829fab8f73ecc033252be31a7cf6/lsp-mode.el#L3009)? I.e. `(accept-process-output nil lsp-response-timeout)`.

File: lsp-mode.el
Comment (yyoncho) [reply lsp-mode.el:5246]: Something like that (clj)

```

(let [even true
      small true]
  (cond->> (range 20)
    even (filter even?)
    small (filter #(< 10 %))))
```

**Comment by yyoncho** [comment]
> As a side note, why not directly wait for `lsp-response-timeout` [here](https://github.com/emacs-lsp/lsp-mode/blob/0349a1cc0976829fab8f73ecc033252be31a7cf6/lsp-mode.el#L3009)? I.e. `(accept-process-output nil lsp-response-timeout)`.

We are not doing accept-process-output on the process that we need to read from and also we need to read multiple messages and potentially from different processes. As a side note, this code should be rewritten to use emacs 26 threads.

**Comment by leungbk** [comment]
> As a side note, this code should be rewritten to use emacs 26 threads.

I've heard the implementation isn't thread-safe:

> ThreadSanitizer (TSan) quickly shows that Emacs’ threading implementation has many data races, making it completely untrustworthy. Until this is fixed, nobody should use Emacs threads for any purpose, and threads should disabled at compile time.

https://nullprogram.com/blog/2018/05/31/

**Comment by yyoncho** [comment]
> > As a side note, this code should be rewritten to use emacs 26 threads.
> 
> I've heard Emacs' threading implementation isn't thread-safe:
> 
> > ThreadSanitizer (TSan) quickly shows that Emacs’ threading implementation has many data races, making it completely untrustworthy. Until this is fixed, nobody should use Emacs threads for any purpose, and threads should disabled at compile time.
> 
> https://nullprogram.com/blog/2018/05/31/

As per Eli's comments, it is not that bad as described in this blogpost.

File: lsp-mode.el
Comment (nbfalcon) [inline lsp-mode.el]: Do we need this commit at all? I don't know how far you want to go to support spec-breaking LS.

Comment (yyoncho) [reply lsp-mode.el]: Sorry, I got lost. Can you elaborate? There are several improvements in this PR so IMHO better to have them in.

Comment (nbfalcon) [reply lsp-mode.el]: Code actions can have a `disabled?` field, which, as per the spec, *must* have a `reason` member. This commit handles the case where language servers send invalid code actions that have a `disabled` field without a `reason` (I don't know of an LS that does that, though). This is just being defensive, and I'm unsure whether we need to handle that case (it is a spec violation after all).

Comment (yyoncho) [reply lsp-mode.el]: Most of the time we don't handle them. Leaving up to you to decide.

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5257]: Is this ok? In case of all actions are disabled, we show all actions instead of empty list?
Also, shouldn't we need guard of `action` as the old code, in case of user not select anything.

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5264]: `-some->>`?

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5285]: `-some->>`?

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:4894]: docstring?

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:4918]: These should be `-find` instead of `-find-index`, the index passing is only working in `ivy-mode`

Comment (nbfalcon) [reply lsp-mode.el:4918]: I have implemented that for `lsp--completing-read`.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5285]: With `->>`, we throw `lsp-no-code-actions`, which is right imho.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5264]: Same as below.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:4895]: Fixed. In what cases do we need docstrings? In all new code, or just in certain functions?

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5258]: The idea is to show an error message in case there is no enabled code action, if the user has enabled `lsp-auto-execute-action` as t ([discussion](https://github.com/emacs-lsp/lsp-mode/pull/2402#discussion_r541268228)). However, I agree that this might be suboptimal. Do you have an idea for a better UI in this case?

As for guarding, we needn't do that, since the user cannot exit from `lsp--select-action` without either selecting an action or `quit`ting (causing a non-local exit), due to REQUIRE-MATCH being set as t.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: I propose to keep it in then, because it can't hurt to have a more lenient lsp implementation.

**Comment by nbfalcon** [comment]
This PR is done. I will not be adding any new commits/features except those needed to address review comments.

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:4914]: Ah, sorry, I missed that.
Also, why don't just pass using `-find` instead of `-find-index`?
It seems that you're taking a round trip of finding index and taking the list element by index here.

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el]: I think we should have docstring in all new function

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:4888]: This is not used?

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:5253]: Isn't `lsp-auto-execute-action` is only for single code action?
The case of `lsp-execute-code-action-by-kind` is falling into the bellow in the spec, should we just take the first code action and let-bind `lsp-auto-execute-action` to `t` so the error will always be shown?
```
	 * - If the user has a keybinding that auto applies a code action and only
	 *   a disabled code actions are returned, the client should show the user
	 *   an error message with `reason` in the editor.
```

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:4914]: If the COLLECTION is a list of strings that is to be transformed, there would be an ambiguitiy: should DEF be used to look up the corresponding string in COLLECTION, or be the DEF argument to completing read?

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: Fixed. I wanted to add a help-echo to disabled code actions at some point, but then realized that it wouldn't work with completing-read.

Comment (nbfalcon) [reply lsp-mode.el]: Nevermind, this is a regression from the commit to support both preferred and disabled code actions at the same time. It was supposed to show the reason. I'll need to add it back.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5255]: We could `let`-bind `lsp-auto-execute-action` to t if there are no enabled ones, however I am unsure if that is the right thing: `lsp-execute-code-action-by-kind` doesn't auto-execute a single code action if available by default, either. If there is only one disabled code action, a prompt containing only it will show up, and along the code action title the `disabled-reason` will be shown:

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: I'll keep that in mind.

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5270]: There's `lsp-make-interactive-code-action` macro which use `lsp-execute-code-action-by-kind`, do you want to provide similar one for `lsp-execute-code-action-by-type`?

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:4916]: Wouldn't that ambiguity be resolved if we document it? For example, DEF should be the corresponding string in COLLECTION (since it's what's returned). That would also be aligned with original `completing-read` which is what wrapped in `lsp--completing-read`.

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:5255]: Shouldn't `lsp--select-action` automatically choose the only one code action available when `lsp-auto-execute-action` is `t`?
Is there special case for disable code action that make the `lsp--select-action` behaves otherwise?

File: lsp-mode.el
Comment (michaelpj) [inline lsp-mode.el:5241]: While we're here... could we make this a list of kinds? The protocol supports that.

File: lsp-mode.el
Comment (michaelpj) [inline lsp-mode.el:5250]:
```suggestion
  "Execute code action by kind."
```

File: lsp-mode.el
Comment (michaelpj) [inline lsp-mode.el:5258]: The difference between this and `lsp-execute-code-action-by-kind` isn't clear to me (or from the docs).

File: lsp-mode.el
Comment (michaelpj) [inline lsp-mode.el:5296]: Should this have an optional `kinds` argument? Generally, it seems like almost all the functions that operate on code actions could take an optional `kinds` argument, perhaps that would let us get away with fewer functions.

Comment (nbfalcon) [reply lsp-mode.el:5296]: Not in this case. `lsp-execute-code-action` should execute whatever action it is given. It is only additionally interactive.

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el:5258]: `lsp-execute-code-action-by-type` matches base-types, i.e. `lsp-execute-code-action-by-type "source"` would match "source.organizeImports", while `-by-kind` matches exactly, i.e. the same example would yield nothing, and it would have to be `(lsp-execute-code-action-by-kind "source.organizeImports")`.

**Review by yyoncho — APPROVED** [review]

File: lsp-mode.el
Comment (michaelpj) [reply lsp-mode.el:5258]: I'm not sure this is a distinction the spec really knows about - certainly "base type" doesn't appear. They do talk about "base kinds", but I'm not sure they're meaningfully distinct from "kinds", it just indicates that they should be hierarchical parents of a set of kinds.

What about just having an `exact-match` argument to `lsp-execute-code-action-by-kind`? Or maybe we shouldn't even provide exact matching. The intention for spec methods that allow you to do kind-based filtering is to include kinds "under" the selected kinds (see e.g. https://github.com/microsoft/language-server-protocol/issues/970#issuecomment-633938017). Do we have a usecase for exact matching?

**Comment by michaelpj** [comment]
It would be nice to get this in. The helpers for executing code actions by kind would be quite nice for clients.

**Pull request closed without merging**
